lib/webfuse/core/dlist.c

```diff
@@ -0,0 +1,101 @@
+#include "webfuse/core/dlist.h"
+#include "webfuse/core/util.h"
+#include <stddef.h>
+
+void wf_dlist_item_init(
+    struct wf_dlist_item * item)
+{
+    item->next = NULL;
+    item->prev = NULL;
+}
+
+void wf_dlist_item_cleanup(
+    struct wf_dlist_item * item)
+{
+    item->next = NULL;
+    item->prev = NULL;
+}
+
+void wf_dlist_init(
+    struct wf_dlist * list)
+{
+    list->first = NULL;
+}
+
+static void wf_dlist_item_cleanup_default(
+    struct wf_dlist_item * WF_UNUSED_PARAM(item),
+    void * WF_UNUSED_PARAM(user_data))
+{
+    // empty
+}
+
+void wf_dlist_cleanup(
+    struct wf_dlist * list,
+    wf_dlist_item_cleanup_fn * cleanup,
+    void * user_data)
+{
+    wf_dlist_item_cleanup_fn * effective_cleanup = (NULL != cleanup) ? cleanup : &wf_dlist_item_cleanup_default;
+    struct wf_dlist_item * item = list->first;
+ 
+    while (NULL != item)
+    {
+        struct wf_dlist_item * next = item->next;
+        effective_cleanup(item, user_data);
+        item = next;
+    }
+
+    list->first = NULL;
+}
+
+void wf_dlist_prepend(
+    struct wf_dlist * list,
+    struct wf_dlist_item * item)
+{
+    if (NULL != list->first)
```

Comment:
If you implement a circular doubly linked lists, you could avoid the unnecessary null pointer checks.
A good example is provided by the linux kernel https://github.com/torvalds/linux/blob/master/include/linux/list.h (But be aware not to copy the implementation :) )
But I think you don't need a doubly linked list at all.

Reply:
Switched to singly linked list.

Since I don't like this macro stuff, I would never think of copiing such stuff ;-)
But thanks for the link.